gcc/config/riscv/peephole_pack.md

```diff
@@ -0,0 +1,587 @@
+ (REGNO (operands[4]) == REGNO (operands[0]) ||
+  peep2_reg_dead_p (3, operands[4]))"
+  [(set (match_dup 0) (plus:SI (match_dup 2) (match_dup 1)))]
+{
+  emit_insn(gen_riscv_bitmanip_packhsi(operands[0], gen_rtx_REG(SImode, REGNO(operands[2])), operands[1])); 
+  DONE;
+})
+(define_peephole2
+  [(set (match_operand:SI 3 "register_operand")
+      (and:SI (ashift:SI (match_operand:SI 1 "register_operand")
+	                     (const_int 8))
+              (const_int 65280)))
+   (set (match_operand:SI 4 "register_operand")
+      (zero_extend:SI (match_operand:QI 2 "register_operand")))
+   (set (match_operand:SI 0 "register_operand")
+      (ior:SI  (match_dup 4)
+			   (match_dup 3)
+ 			   ))]
+ "TARGET_ZBP &&
+ !TARGET_64BIT &&
+ (REGNO (operands[3]) == REGNO (operands[0]) ||
+  peep2_reg_dead_p (3, operands[3])) &&
+ (REGNO (operands[4]) == REGNO (operands[0]) ||
+  peep2_reg_dead_p (3, operands[4]))"
+  [(set (match_dup 0) (plus:SI (match_dup 2) (match_dup 1)))]
+{
+  emit_insn(gen_riscv_bitmanip_packhsi(operands[0], gen_rtx_REG(SImode, REGNO(operands[2])), operands[1])); 
+  DONE;
+})
+
+
+(define_peephole2
+  [(set (match_operand:SI 3 "register_operand")
+        (zero_extend:SI (match_operand:QI 1 "nonimmediate_operand")))
+   (set (match_operand:SI 4 "register_operand")
+        (zero_extend:SI (match_operand:QI 2 "nonimmediate_operand")))
+   (set (match_operand:SI 5 "register_operand")
+        (ashift:SI (match_dup 3)
+	               (const_int 8)))
+   (set (match_operand:SI 0 "register_operand")
+      (ior:SI  (match_dup 5)
+			   (match_dup 4)
+ 			   ))]
+ "TARGET_ZBP &&
+ !TARGET_64BIT &&
+ (REGNO (operands[5]) == REGNO (operands[0]) ||
+  peep2_reg_dead_p (4, operands[5]))"
+  [(set (match_dup 0) (plus:SI (match_dup 2) (match_dup 1)))]
+{
+  emit_insn(gen_zero_extendqisi2(operands[3], operands[1]));
+  emit_insn(gen_zero_extendqisi2(operands[4], operands[2]));
+  emit_insn(gen_riscv_bitmanip_packhsi(operands[0], operands[4], operands[3])); 
+  DONE;
+})
+(define_peephole2
+  [(set (match_operand:SI 3 "register_operand")
+        (zero_extend:SI (match_operand:QI 1 "nonimmediate_operand")))
+   (set (match_operand:SI 4 "register_operand")
+        (zero_extend:SI (match_operand:QI 2 "nonimmediate_operand")))
+   (set (match_operand:SI 5 "register_operand")
+        (ashift:SI (match_dup 3)
+	               (const_int 8)))
+   (set (match_operand:SI 0 "register_operand")
+      (ior:SI  (match_dup 4)
+			   (match_dup 5)
+ 			   ))]
+ "TARGET_ZBP &&
+ !TARGET_64BIT &&
+ (REGNO (operands[5]) == REGNO (operands[0]) ||
+  peep2_reg_dead_p (4, operands[5]))"
+  [(set (match_dup 0) (plus:SI (match_dup 2) (match_dup 1)))]
+{
+  emit_insn(gen_zero_extendqisi2(operands[3], operands[1]));
+  emit_insn(gen_zero_extendqisi2(operands[4], operands[2]));
+  emit_insn(gen_riscv_bitmanip_packhsi(operands[0], operands[4], operands[3])); 
+  DONE;
+})
+
+;; Does it look ad-hoc? I think it looks ad-hoc...
+(define_peephole2
```

Comment:
I believe those ad-hoc pattern could be removed once `pack*` pattern has describe right :)
Or...at least could be simplified.

Reply:
Actually they still seem necessary...